sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/WriteToDestinations.java

```diff
@@ -66,7 +61,58 @@ class WriteToDestinations extends PTransform<PCollection<Row>, IcebergWriteResul
   }
 
   @Override
-  public IcebergWriteResult expand(PCollection<Row> input) {
+  public IcebergWriteResult expand(PCollection<KV<String, Row>> input) {
+    // Write records to files
+    PCollection<FileWriteResult> writtenFiles =
+        input.isBounded().equals(PCollection.IsBounded.UNBOUNDED)
+            ? writeTriggered(input)
+            : writeUntriggered(input);
+
+    // Commit files to tables
+    PCollection<KV<String, SnapshotInfo>> snapshots =
+        writtenFiles.apply(new AppendFilesToTables(catalogConfig));
+
+    return new IcebergWriteResult(input.getPipeline(), snapshots);
+  }
+
+  private PCollection<FileWriteResult> writeTriggered(PCollection<KV<String, Row>> input) {
+    checkArgumentNotNull(
```

Comment:
Why was this removed from the IO, and moved into the transform? Wouldn't it have been better in the IO layer (we could fail earlier?)

Reply:
I moved it here because I thought it'd be cleaner to have all the streaming-related code in one place

We will fail early either way, this check still happens during pipeline construction time.